File: app/src/androidTest/java/io/catter2/favorites_activity/FavoritesActivityTest.java
package io.catter2.favorites_activity;

import android.content.Intent;
import android.support.test.espresso.matcher.ViewMatchers;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.List;

import io.catter2.R;
import io.catter2.di.FavoritesRepoDIModule;
import io.catter2.di.UserDIComponent;
import io.catter2.espresso_utils.RecyclerViewItemCountAssertion;
import io.catter2.favorites.FavoritesRepository;
import io.catter2.favorites.GetFavoritesUseCase;

import static android.support.test.espresso.Espresso.onView;
import static org.mockito.Mockito.mock;

@RunWith(AndroidJUnit4.class)
public class FavoritesActivityTest {
    @Rule
    public ActivityTestRule<FavoritesActivity> activityRule =
            new ActivityTestRule<>(FavoritesActivity.class, true, false /* Do not launch activity! */);

    @BeforeClass
    public static void beforeClass() {
        // Warning: The AppDIComponent is still being created by the 'App' class!
        UserDIComponent.initialize(mock(FavoritesRepoDIModule.class));
    }

    @Test
    public void testThereAreThreeImages() {
        final ArrayList<String> expectedUrls = new ArrayList<>();
        expectedUrls.add("url0");
        expectedUrls.add("url1");
        expectedUrls.add("url2");
        setUpGetFavoritesResponse(expectedUrls);

        activityRule.launchActivity(new Intent());

        onView(ViewMatchers.withId(R.id.favorites_rv)).check(new RecyclerViewItemCountAssertion(3));
    }

    @Test
    public void testThereAreZeroImages() {
        final ArrayList<String> expectedUrls = new ArrayList<>();
        setUpGetFavoritesResponse(expectedUrls);

        activityRule.launchActivity(new Intent());

        onView(ViewMatchers.withId(R.id.favorites_rv)).check(new RecyclerViewItemCountAssertion(0));
    }

    private void setUpGetFavoritesResponse(final List<String> urls) {
        FavoritesActivityDIModule.testGetFavoritesUseCase =
                new GetFavoritesUseCase(mock(FavoritesRepository.class)) {
                    @Override
                    public void getFavorites(Callback callback) {
                        callback.favoriteUrlsUpdated(urls);
                    }
                };
    }
}

File: app/src/androidTest/java/io/catter2/favorites_activity/FavoritesActivityTest.kt
package io.catter2.favorites_activity

import android.content.Intent
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.matcher.ViewMatchers
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import com.nhaarman.mockito_kotlin.mock
import io.catter2.R
import io.catter2.di.UserKodein
import io.catter2.espresso_utils.RecyclerViewItemCountAssertion
import io.catter2.favorites.GetFavoritesUseCase
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

@RunWith(AndroidJUnit4::class)
open class FavoritesActivityTest {
    companion object {
        @BeforeClass
        @JvmStatic
        fun beforeClass() {
            // Warning: The AppDIComponent is still being created by the 'App' class!
            UserKodein.initialize("user")
        }
    }

    @get:Rule
    var activityRule =
            ActivityTestRule(FavoritesActivity::class.java, true, false /* Do not launch activity! */)

    @Test
    fun testThereAreThreeImages() {
        val expectedUrls = ArrayList<String>()
        expectedUrls.add("url0")
        expectedUrls.add("url1")
        expectedUrls.add("url2")
        setUpGetFavoritesResponse(expectedUrls)

        activityRule.launchActivity(Intent())

        onView(ViewMatchers.withId(R.id.favorites_rv)).check(RecyclerViewItemCountAssertion(3))
    }

    @Test
    fun testThereAreZeroImages() {
        val expectedUrls = ArrayList<String>()
        setUpGetFavoritesResponse(expectedUrls)

        activityRule.launchActivity(Intent())

        onView(ViewMatchers.withId(R.id.favorites_rv)).check(RecyclerViewItemCountAssertion(0))
    }

    private fun setUpGetFavoritesResponse(urls: List<String>) {
        FavoritesActivityKodein.testGetFavoritesUseCase =
                object : GetFavoritesUseCase(mock()) {
                    override fun getFavorites(callback: (List<String>) -> Unit) {
                        callback(urls)
                    }
                }
    }
}
